pkg/serving/config_changes.go

```diff
@@ -304,3 +425,255 @@ func removeEnvVars(env []corev1.EnvVar, toRemove []string) []corev1.EnvVar {
+				break
+			}
+		}
+	}
+
+	if len(volumes) == 0 {
+		return nil, nil
+	}
+
+	return volumes, nil
+}
+
+// =======================================================================================
+
+type volumeSourceInfo struct {
+	volumeSourceType VolumeSourceType
+	volumeSourceName string
+}
+
+func newVolumeSourceInfoWithSpecString(spec string) (*volumeSourceInfo, error) {
+	slices := strings.SplitN(spec, ":", 2)
+	if len(slices) != 2 {
+		return nil, fmt.Errorf("argument requires a value that contains the : character; got %q", spec)
+	}
+
+	var volumeSourceType VolumeSourceType
+
+	typeString := strings.TrimSpace(slices[0])
+	volumeSourceName := strings.TrimSpace(slices[1])
+
+	switch typeString {
+	case "config-map", "cm":
+		volumeSourceType = ConfigMapVolumeSourceType
+	case "secret", "sc":
+		volumeSourceType = SecretVolumeSourceType
+	default:
+		return nil, fmt.Errorf("unsupported volume source type \"%q\"; supported volume source types are \"config-map\" and \"secret\"", slices[0])
+	}
+
+	if len(volumeSourceName) == 0 {
+		return nil, fmt.Errorf("the name of %s cannot be an empty string", volumeSourceType)
+	}
+
+	return &volumeSourceInfo{
+		volumeSourceType: volumeSourceType,
+		volumeSourceName: volumeSourceName,
+	}, nil
+}
+
+func (vol *volumeSourceInfo) getCanonicalName() string {
+	return fmt.Sprintf("%s:%s", vol.volumeSourceType, vol.volumeSourceName)
+}
+
+func getCanonicalNameFromEnvFromSource(envSrc *corev1.EnvFromSource) (string, error) {
+	if envSrc.ConfigMapRef != nil {
+		return fmt.Sprintf("%s:%s", ConfigMapVolumeSourceType, envSrc.ConfigMapRef.Name), nil
+	}
+	if envSrc.SecretRef != nil {
+		return fmt.Sprintf("%s:%s", SecretVolumeSourceType, envSrc.SecretRef.Name), nil
+	}
+
+	return "", fmt.Errorf("there is no ConfigMapRef or SecretRef in a EnvFromSource")
+}
+
+func (vol *volumeSourceInfo) createEnvFromSource() *corev1.EnvFromSource {
+	switch vol.volumeSourceType {
+	case ConfigMapVolumeSourceType:
+		return &corev1.EnvFromSource{
+			ConfigMapRef: &corev1.ConfigMapEnvSource{
+				LocalObjectReference: corev1.LocalObjectReference{
+					Name: vol.volumeSourceName,
+				}}}
+	case SecretVolumeSourceType:
+		return &corev1.EnvFromSource{
+			SecretRef: &corev1.SecretEnvSource{
+				LocalObjectReference: corev1.LocalObjectReference{
+					Name: vol.volumeSourceName,
+				}}}
+	}
+
+	return nil
+}
+
+// =======================================================================================
+
+func existsVolumeNameInVolumes(volumeName string, volumes []corev1.Volume) bool {
+	for _, volume := range volumes {
+		if volume.Name == volumeName {
+			return true
+		}
+	}
+	return false
+}
+
+func existsVolumeNameInVolumeMounts(volumeName string, volumeMounts []corev1.VolumeMount) bool {
+	for _, volumeMount := range volumeMounts {
+		if volumeMount.Name == volumeName {
+			return true
+		}
+	}
+	return false
+}
+
+func generateVolumeName(path string) string {
+	return fmt.Sprintf("kn-managed-volume:%s", path)
```

Comment:
This is too simplistic, as you have to sanitize the path that it's a well-formed kubernetes identifier. As the path has to start with `/` (as you can only mount absolute paths) you will always get an invalid name here. Unfortunately, such things can be only detected in integration tests, not unit tests, as unit tests don't do any validation.

Also, I would either don't use a prefix or, if required e.g. when for auto-deleting volumes, then just use `kn:`, which is just shorter and easier to read when examing the k8s files.

Reply:
I've changed the part to use a sanitized path string + checksum of original path string without any prefix.

Reply:
Thanks, works for me now !

I'm going to go over the rest of the PR quickly and lets get it merged today so that we get it into 0.10.0 (to be released today). We can then fix any outstanding issues (or add some real integrations tests) later.